### PR TITLE
hv: vtd: fix potential buffer overflow in suspend/resume

### PR DESCRIPTION
**Pull request by binbinwu1**
In current code of suspend_iommu/resume_iommu, there is potential buffer overflow
according to the code.
This patch put the buffer to struct dmar_drhd_rt, so that no need to access the buffer
via index.

Signed-off-by: Binbin Wu <binbin.wu@intel.com>
Tracked-On: #1252
Acked-by: Eddie Dong <eddie.dong@intel.com>

### CODE REVIEW

File: hypervisor/arch/x86/vtd.c [hunk 1]
```diff
@@ -1230,21 +1230,12 @@ void suspend_iommu(void)
...
 		dmar_disable_translation(dmar_unit);
-
-		/* If the number of real iommu devices is larger than we
-		 * defined in kconfig.
-		 */
-		if (iommu_idx > CONFIG_MAX_IOMMU_NUM) {
-			pr_err("iommu dev number is larger than pre-defined");
-			break;
-		}
-		iommu_idx++;
```
Comment (dongyaozu) [inline hypervisor/arch/x86/vtd.c]: should we remove definition of iommu_idx?

File: hypervisor/arch/x86/vtd.c [hunk 2]
```diff
@@ -30,6 +30,10 @@
 #define ROOT_ENTRY_LOWER_CTP_POS            (12U)
 #define ROOT_ENTRY_LOWER_CTP_MASK           (0xFFFFFFFFFFFFFUL)
 
+/* 4 iommu fault register state */
+#define	IOMMU_FAULT_REGISTER_STATE_NUM	4U
+#define	IOMMU_FAULT_REGISTER_STATE_SIZE	4U
```
Comment (dongyaozu) [inline hypervisor/arch/x86/vtd.c]: This MACRO is actually  sizeof(failt_state), can we do in this way?
It helps to couple them together.

File: hypervisor/arch/x86/vtd.c (hunk 1 above)
Comment (binbinwu1) [reply hypervisor/arch/x86/vtd.c]: will fix it.

File: hypervisor/arch/x86/vtd.c (hunk 2 above)
Comment (binbinwu1) [reply hypervisor/arch/x86/vtd.c]: It is the register width/register address offset of these fault registers. And it is fixed according to vtd spec. sizeof(fault_state) kind of related the hardware register offset to the data structure we defined, which I not prefer to. If we want to get rid of a new macro, I would prefer to use sizeof (uint32_t).

File: hypervisor/arch/x86/vtd.c [hunk 3]
```diff
@@ -1230,29 +1229,20 @@ void suspend_iommu(void)
 
 		/* save IOMMU fault register state */
 		for (i = 0U; i < IOMMU_FAULT_REGISTER_STATE_NUM; i++) {
-			iommu_fault_state[iommu_idx][i] =
-				iommu_read32(dmar_unit, DMAR_FECTL_REG +
-					(i * IOMMU_FAULT_REGISTER_STATE_NUM));
+			dmar_unit->fault_state[i] =  iommu_read32(dmar_unit,
+				DMAR_FECTL_REG + (i * sizeof(uint32_t)));
```
Comment (dongyaozu) [inline hypervisor/arch/x86/vtd.c]: sizeof(uint32_t) is not good --- if the definition side changes, it still lead to bug. should be sizeof(fault_state).
for example  sizeof(  ((struct dmar_drhd_rt*) NULL) ->fault_state[0]).

Comment (binbinwu1) [reply hypervisor/arch/x86/vtd.c]: This value is not related to definition side of fault_state, it should be aligned with hardware spec as below, so it should be a constant value:
#define DMAR_FSTS_REG   0x34U    /* Fault Status register */
#define DMAR_FECTL_REG  0x38U    /* Fault control register */
#define DMAR_FEDATA_REG 0x3cU    /* Fault event interrupt data register */
#define DMAR_FEADDR_REG 0x40U    /* Fault event interrupt addr register */

Comment (dongyaozu) [reply hypervisor/arch/x86/vtd.c]: Spec is the source of the requirement. data structure definition is following the spec.
The MACRO could derive from the data structure or from the original spec. Either way is better than sizeof(uint32_t) here, which follows neither of the source.

Comment (binbinwu1) [reply hypervisor/arch/x86/vtd.c]: fixed

**Comment by haoweiah** [comment]
Ready to merge

**Pull request merged**